python/ray/tune/execution/ray_trial_executor.py

```diff
@@ -497,12 +533,90 @@ def _start_trial(self, trial: Trial) -> bool:
         self.restore(trial)
         self.set_status(trial, Trial.RUNNING)
 
-        self._staged_trials.discard(trial)
+        self._unstage_trial_with_resources(trial)
 
         if not trial.is_restoring:
             self._train(trial)
         return True
 
+    def _unstage_trial_with_resources(self, trial: Trial):
+        # Case 1: The trial we started was staged. Just remove it
+        if trial in self._staged_trials:
+            self._staged_trials.remove(trial)
+            return
+
+        # Case 2: We staged a trial "A" with the same resources, but our trial "B"
+        # was selected by the scheduler to run. The resource manager does not care
+        # about "trials", it just cares about resources being available. Thus we
+        # look for a staged trial with the same resource requirements and remove it
+
+        resource_request = trial.placement_group_factory
+        # Remove staged trial with same resource requirements
+        candidate_trial = None
+        for staged_trial in self._staged_trials:
+            staged_resources = staged_trial.placement_group_factory
+            if staged_resources == resource_request:
+                candidate_trial = staged_trial
+                break
+
+        if candidate_trial:
+            self._staged_trials.remove(candidate_trial)
+            return
+
+        raise RuntimeError(
+            "Started a trial with resources requested by a different trial, but "
+            "this trial was lost. This is an error in Ray Tune's execution "
+            "logic. Please raise a GitHub issue at "
+            "https://github.com/ray-project/ray/issues"
+        )
+
+    def _maybe_cache_trial_actor(self, trial: Trial) -> bool:
+        if not self._reuse_actors:
+            return False
+
+        if (
```

Comment:
For this equation, should we also consider the number of requests that have been filed to ResourceManager?

Reply:
As discussed offline, it's the callers responsibility to keep track, so we should contain the logic there